gno.land/pkg/gnoland/app.go

```diff
@@ -196,9 +225,109 @@ func InitChainer(baseApp *sdk.BaseApp, acctKpr auth.AccountKeeperI, bankKpr bank
 	}
 }
 
-// XXX not used yet.
-func EndBlocker(vmk vm.VMKeeperI) func(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
-	return func(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
-		return abci.ResponseEndBlock{}
+// endBlockerApp is the app abstraction required by any EndBlocker
+type endBlockerApp interface {
+	// LastBlockHeight returns the latest app height
+	LastBlockHeight() int64
+
+	// Logger returns the logger reference
+	Logger() *slog.Logger
+}
+
+// EndBlocker defines the logic executed after every block.
+// Currently, it parses events that happened during execution to calculate
+// validator set changes
+func EndBlocker(
+	collector *collector[validatorUpdate],
+	vmKeeper vm.VMKeeperI,
+	app endBlockerApp,
+) func(
+	ctx sdk.Context,
+	req abci.RequestEndBlock,
+) abci.ResponseEndBlock {
+	return func(ctx sdk.Context, _ abci.RequestEndBlock) abci.ResponseEndBlock {
+		// Check if there was a valset change
+		if len(collector.getEvents()) == 0 {
+			// No valset updates
+			return abci.ResponseEndBlock{}
+		}
+
+		// Run the VM to get the updates from the chain
+		msg := vm.MsgCall{
+			Caller:  crypto.Address{}, // Zero address
+			PkgPath: valRealm,
+			Func:    valChangesFn,
+			Args:    []string{fmt.Sprintf("%d", app.LastBlockHeight())},
+		}
+
+		response, err := vmKeeper.Call(ctx, msg)
+		if err != nil {
+			app.Logger().Error("unable to call VM during EndBlocker", "err", err)
+
+			return abci.ResponseEndBlock{}
+		}
+
+		// Extract the updates from the VM response
+		updates, err := extractUpdatesFromResponse(response)
+		if err != nil {
+			app.Logger().Error("unable to extract updates from response", "err", err)
+
+			return abci.ResponseEndBlock{}
+		}
+
+		return abci.ResponseEndBlock{
+			ValidatorUpdates: updates,
+		}
+	}
+}
+
+// extractUpdatesFromResponse extracts the validator set updates
+// from the VM response.
+//
+// This method is not ideal, but currently there is no mechanism
+// in place to parse typed VM responses
+func extractUpdatesFromResponse(response string) ([]abci.ValidatorUpdate, error) {
+	// Find the submatches
+	matches := valRegexp.FindAllStringSubmatch(response, -1)
+	if len(matches) == 0 {
+		// No changes to extract
+		return nil, nil
+	}
+
+	updates := make([]abci.ValidatorUpdate, 0, len(matches))
```

Comment:
```suggestion
	updates := make([]abci.ValidatorUpdate, 0)
```

you're using append.

Reply:
Yes, but I know exactly how many elements I have in the updates `len(matches)` -- I'm specifying the capacity for the slice